File: src/services/DatabaseService.ts
import { initializeDatabase, testConnection, closeDatabase } from '../config/database';
import { PatientService } from './PatientService';
import { OxygenService } from './OxygenService';

// Re-export all services for easy importing
export { PatientService } from './PatientService';
export { OxygenService } from './OxygenService';

/**
 * Main Database Service
 * Central hub for all database operations and services
 */
export class DatabaseService {
  private static initialized = false;

  /**
   * Initialize the database connection and all services
   */
  static async initialize(): Promise<boolean> {
    try {
      if (this.initialized) {
        console.log('Database already initialized');
        return true;
      }

      console.log('Initializing database connection...');
      
      // Initialize connection pool
      initializeDatabase();
      
      // Test the connection
      const isConnected = await testConnection();
      
      if (!isConnected) {
        throw new Error('Failed to connect to database');
      }

      this.initialized = true;
      console.log('✅ Database service initialized successfully');
      
      return true;
    } catch (error) {
      console.error('❌ Failed to initialize database service:', error);
      return false;
    }
  }

  /**
   * Close all database connections
   */
  static async shutdown(): Promise<void> {
    try {
      await closeDatabase();
      this.initialized = false;
      console.log('✅ Database service shut down successfully');
    } catch (error) {
      console.error('❌ Error during database shutdown:', error);
    }
  }

  /**
   * Check if database service is initialized
   */
  static isInitialized(): boolean {
    return this.initialized;
  }

  /**
   * Test database connectivity
   */
  static async healthCheck(): Promise<{
    status: 'healthy' | 'unhealthy';
    message: string;
    timestamp: Date;
  }> {
    try {
      const isHealthy = await testConnection();
      
      return {
        status: isHealthy ? 'healthy' : 'unhealthy',
        message: isHealthy ? 'Database connection is working' : 'Database connection failed',
        timestamp: new Date()
      };
    } catch (error) {
      return {
        status: 'unhealthy',
        message: error instanceof Error ? error.message : 'Unknown database error',
        timestamp: new Date()
      };
    }
  }

  /**
   * Get database statistics
   */
  static async getSystemStats(): Promise<{
    patients: any;
    oxygen: any;
    timestamp: Date;
  }> {
    try {
      const [patientStats, oxygenStats] = await Promise.all([
        PatientService.getStatistics(),
        OxygenService.getStatistics()
      ]);

      return {
        patients: patientStats.success ? patientStats.data : null,
        oxygen: oxygenStats.success ? oxygenStats.data : null,
        timestamp: new Date()
      };
    } catch (error) {
      console.error('Error getting system stats:', error);
      return {
        patients: null,
        oxygen: null,
        timestamp: new Date()
      };
    }
  }
}

// Graceful shutdown handler
process.on('SIGINT', async () => {
  console.log('Received SIGINT, shutting down database service...');
  await DatabaseService.shutdown();
  process.exit(0);
});

process.on('SIGTERM', async () => {
  console.log('Received SIGTERM, shutting down database service...');
  await DatabaseService.shutdown();
  process.exit(0);
});

export default DatabaseService;

File: backend/services/database_service.py
import logging
from datetime import datetime
from config.database import initialize_database, test_connection, close_database
from services.patient_service import PatientService
from services.oxygen_service import OxygenService

logger = logging.getLogger(__name__)

class DatabaseService:
    """Main Database Service - Central hub for all database operations and services"""
    
    _initialized = False
    
    @classmethod
    def initialize(cls) -> bool:
        """Initialize the database connection and all services"""
        try:
            if cls._initialized:
                logger.info('Database already initialized')
                return True
            
            logger.info('Initializing database connection...')
            
            # Initialize connection pool
            initialize_database()
            
            # Test the connection
            is_connected = test_connection()
            
            if not is_connected:
                raise Exception('Failed to connect to database')
            
            cls._initialized = True
            logger.info('✅ Database service initialized successfully')
            
            return True
            
        except Exception as e:
            logger.error(f'❌ Failed to initialize database service: {e}')
            return False
    
    @classmethod
    def shutdown(cls):
        """Close all database connections"""
        try:
            close_database()
            cls._initialized = False
            logger.info('✅ Database service shut down successfully')
            
        except Exception as e:
            logger.error(f'❌ Error during database shutdown: {e}')
    
    @classmethod
    def is_initialized(cls) -> bool:
        """Check if database service is initialized"""
        return cls._initialized
    
    @classmethod
    def health_check(cls) -> dict:
        """Test database connectivity"""
        try:
            is_healthy = test_connection()
            
            return {
                'status': 'healthy' if is_healthy else 'unhealthy',
                'message': 'Database connection is working' if is_healthy else 'Database connection failed',
                'timestamp': str(datetime.now())
            }
            
        except Exception as e:
            return {
                'status': 'unhealthy',
                'message': str(e),
                'timestamp': str(datetime.now())
            }
    
    @classmethod
    def get_system_stats(cls) -> dict:
        """Get database statistics"""
        try:
            patient_stats = PatientService.get_statistics()
            oxygen_stats = OxygenService.get_statistics()
            
            return {
                'patients': patient_stats.data if patient_stats.success else None,
                'oxygen': oxygen_stats.data if oxygen_stats.success else None,
                'timestamp': str(datetime.now())
            }
            
        except Exception as e:
            logger.error(f'Error getting system stats: {e}')
            return {
                'patients': None,
                'oxygen': None,
                'timestamp': str(datetime.now())
            }